## ¿Cual es la probabilidad de que se forme un bono con tiempo de madurez  mayor a 10 y precio mayor a 105?

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

b = pd.read_csv('bonds prices.csv')
b.head()

id  bond_id  trade_price     weight  current_coupon  time_to_maturity  \
0   1        1      128.596  11.316404            5.95         26.031126   
1   2        1      126.673   8.763446            5.95         26.016697   
2   3        1      126.880   6.499730            5.95         26.008760   
3   4        1      126.362   0.865854            5.95         26.008619   
4   5        1      124.918  11.938543            5.95         25.981840   

   is_callable  reporting_delay  trade_size  trade_type  \
0            0            1.731      120000           2   
1            0           30.518       85000           4   
2            0            4.424     4200000           3   
3            0           43.059     5000001           2   
4            0            6.332       10000           2   

            ...             received_time_diff_last1  trade_price_last1  \
0           ...                               758762            129.089   
1           ...                               455028            128.596   
2           ...                               250310            126.673   
3           ...                                 4441            126.880   
4           ...                               844484            126.362   

   trade_size_last1  trade_type_last1  curve_based_price_last1  \
0            200000                 2               129.435587   
1            120000                 2               127.561201   
2             85000                 4               126.323528   
3           4200000                 3               126.761433   
4           5000001                 2               127.025765   

   received_time_diff_last2  trade_price_last2  trade_size_last2  \
0                       NaN                NaN               NaN   
1                 1213791.0            129.089          200000.0   
2                  705339.0            128.596          120000.0   
3                  254752.0            126.673           85000.0   
4                  848925.0            126.880         4200000.0   

   trade_type_last2  curve_based_price_last2  
0               NaN                      NaN  
1               2.0               129.435587  
2               2.0               127.561201  
3               4.0               126.323528  
4               3.0               126.761433  

[5 rows x 21 columns]

In [ ]:
sns.boxplot(b.trade_price)
sns.boxplot(b.time_to_maturity)

In [2]:
b.trade_price.describe()

count    762678.000000
mean        103.440546
std           9.824289
min           1.009000
25%          99.433000
50%         102.300000
75%         107.110000
max         168.148000
Name: trade_price, dtype: float64

In [3]:
b.time_to_maturity.describe()

count    762678.000000
mean          6.564305
std           6.267664
min           0.082471
25%           2.493513
50%           4.775924
75%           8.258087
max          84.159498
Name: time_to_maturity, dtype: float64

In [2]:
posibles = b.trade_price[(b.trade_price > 105)&(b.time_to_maturity > 10)].count()
posibles
p1 = float(posibles)/float(762678)
p1

0.05137817007964043

In [11]:
pd.crosstab((b.trade_price > 105), (b.time_to_maturity > 10))

time_to_maturity   False  True 
trade_price                    
False             453023  55326
True              215144  39185

In [12]:
b.groupby(['time_to_maturity','trade_price']).count()

id  bond_id  weight  current_coupon  \
time_to_maturity trade_price                                        
0.082471         100.260000    1        1       1               1   
0.082472         100.260000    1        1       1               1   
0.082737         100.250000    1        1       1               1   
                 100.250000    1        1       1               1   
                 100.252000    1        1       1               1   
0.083092         100.275000    1        1       1               1   
0.085313         100.272000    1        1       1               1   
                 100.272000    1        1       1               1   
0.085406         100.272000    1        1       1               1   
                 100.272000    1        1       1               1   
0.090999         100.321917    1        1       1               1   
0.091000         100.321917    1        1       1               1   
0.091002         100.321917    1        1       1               1   
0.091004         100.321917    1        1       1               1   
0.091005         100.321917    1        1       1               1   
0.091007         100.321917    1        1       1               1   
0.091010         100.321917    1        1       1               1   
0.091011         100.321917    1        1       1               1   
0.091017         100.321917    1        1       1               1   
0.091427         100.410000    1        1       1               1   
0.091437         100.222000    1        1       1               1   
                 100.222000    1        1       1               1   
                 100.124500    1        1       1               1   
0.091557         100.410000    1        1       1               1   
0.096369         100.301000    1        1       1               1   
                 100.301000    1        1       1               1   
0.096761         100.000000    2        2       2               2   
0.098883         100.507000    1        1       1               1   
0.098938         100.413000    1        1       1               1   
0.098961         100.489734    1        1       1               1   
...                           ..      ...     ...             ...   
84.080222        117.250000    1        1       1               1   
84.080225        118.000000    1        1       1               1   
84.080254        115.060000    2        2       2               2   
84.080377        116.250000    1        1       1               1   
84.080378        116.187500    1        1       1               1   
84.080438        114.860000    1        1       1               1   
                 115.360000    1        1       1               1   
84.099600        122.000000    1        1       1               1   
                 122.000000    1        1       1               1   
84.099623        121.950000    1        1       1               1   
84.099628        121.950000    1        1       1               1   
84.115973        122.621000    1        1       1               1   
84.115974        122.219000    1        1       1               1   
84.118683        115.443000    1        1       1               1   
                 114.866000    1        1       1               1   
84.118836        117.340000    1        1       1               1   
                 115.840000    1        1       1               1   
84.118847        120.127200    1        1       1               1   
84.118868        120.741000    1        1       1               1   
84.118890        120.783000    1        1       1               1   
84.120989        115.840000    1        1       1               1   
84.129590        120.700000    1        1       1               1   
84.129593        119.200000    1        1       1               1   
84.140084        117.218000    1        1       1               1   
                 116.632000    1        1       1               1   
84.140650        115.100000 

### R. La probabilidad de obtener un bono mayor a 10 años y con precio mayor a 105, alrededor de el ultio cuartil, es = 0.05137817007964043 es decir me trae cuenta a largo plazo ya que no hay tanta demanda.

## ¿Cual es la probabilidad que un bono de alto o bajo interes pueda ser liquidado en cualquier momento?

In [4]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

b = pd.read_csv('bonds prices.csv')

In [ ]:
sns.boxplot(b.current_coupon)
sns.boxplot(b.trade_price)

In [5]:
b.current_coupon.describe()

count    762678.000000
mean          5.588142
std           1.724374
min           0.000000
25%           4.750000
50%           5.625000
75%           6.500000
max          13.500000
Name: current_coupon, dtype: float64

In [6]:
b.trade_price.describe()

count    762678.000000
mean        103.440546
std           9.824289
min           1.009000
25%          99.433000
50%         102.300000
75%         107.110000
max         168.148000
Name: trade_price, dtype: float64

In [14]:
pd.crosstab((b.current_coupon > 6.5), (b.is_callable == 1))

is_callable      False  True 
current_coupon               
False           554726  32936
True            122986  52030

In [9]:
def coupon_callable(couponx):
    are_callable = b.is_callable[(b.is_callable == 1)]
    mayores = are_callable[(b.current_coupon >= 6.5)]
    menores = are_callable[(b.current_coupon < 6.5)]
    if couponx >= 6.5:
        x = float(mayores.count())/float(are_callable.count())
    else:
        x = float(menores.count())/float(are_callable.count())
    return x


coupon_callable(8) #0.6401854859590895
coupon_callable(3) #0.35981451404091047

0.35981451404091047

In [13]:
b.groupby(['current_coupon','trade_price']).count()

id  bond_id  weight  time_to_maturity  \
current_coupon trade_price                                          
0.0            55.745900     1        1       1                 1   
               56.010000     1        1       1                 1   
               62.420000     1        1       1                 1   
               63.640500     1        1       1                 1   
               64.000000     1        1       1                 1   
               64.500000     1        1       1                 1   
               64.597000     1        1       1                 1   
               64.750000     2        2       2                 2   
               64.851000     1        1       1                 1   
               64.860000     1        1       1                 1   
               64.964437     1        1       1                 1   
               65.000000     3        3       3                 3   
               65.040000     1        1       1                 1   
               65.244600     1        1       1                 1   
               65.327000     1        1       1                 1   
               65.382000     1        1       1                 1   
               65.400000     1        1       1                 1   
               65.420000     1        1       1                 1   
               65.454000     1        1       1                 1   
               65.510000     1        1       1                 1   
               65.530000     1        1       1                 1   
               65.560000     1        1       1                 1   
               65.574000     1        1       1                 1   
               65.576000     1        1       1                 1   
               65.655000     1        1       1                 1   
               65.750000     2        2       2                 2   
               65.774000     1        1       1                 1   
               65.825000     2        2       2                 2   
               65.835000     2        2       2                 2   
               65.882000     3        3       3                 3   
...                         ..      ...     ...               ...   
13.5           114.000000    7        7       7                 7   
               114.062500    2        2       2                 2   
               114.250000    5        5       5                 5   
               114.375000    1        1       1                 1   
               114.500000    4        4       4                 4   
               114.625000    1        1       1                 1   
               114.750000    4        4       4                 4   
               114.812500    2        2       2                 2   
               114.875000    1        1       1                 1   
               115.000000   16       16      16                16   
               115.050000    1        1       1                 1   
               115.125000    4        4       4                 4   
               115.150000    1        1       1                 1   
               115.220000    1        1       1                 1   
               115.237500    2        2       2                 2   
               115.250000   20       20      20                20   
               115.262500    2        2       2                 2   
               115.312500    3        3       3                 3   
               115.375000   11       11      11                11   
               115.400000    1        1       1                 1   
               115.437500    6        6       6                 6   
               115.500000   17       17      17                17   
               115.520000    1        1       1                 1   
               115.562500    4        4       4                 4   
               115.625000    2        2       2                 2   
               115.700000 

## Probabilidad de un bono alto es 0.6401854859590895
## Probabilidad de un bono bajo es 0.35981451404091047